### ISSUE TITLE
NVDA says that our sliders have "invalid entry"

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/sun#413]
Found by @KatieWoe for https://github.com/phetsims/QA/issues/219. For some of our sliders, NVDA says that values have "invalid entries". I would guess that it is caused by this line:

```js
// HTML requires that the value be evenly divisible by the step size to receive 'input' events, which
// is critical to work with mobile AT like VoiceOver
self.setAccessibleAttribute( 'step', ( enabledRange.max - enabledRange.min ) / 100 );
```

**Comment by jessegreenberg** [comment]
Fixed with the above commit. Leaving open as pending-testing because this came from https://github.com/phetsims/QA/issues/219

**Comment by jessegreenberg** [comment]
@mbarlow12 showed me that this change broke sliders in Safari when VO is enabled, we need a different fix.

**Comment by mbarlow12** [comment]
In GFL & RIAW, we found that the aria-value text wasn't being read out. There's also some odd behavior when using the VO keys (ctrl-opt on Mac) within the slider to increment the value. In GFL for example, moving the position sliders in this way, you're unable to increment the value past 4.6 on the left object.

**Comment by jessegreenberg** [comment]
Two issues with the current implementation, both showing up in cases we are hitting in GFL. We are allowing negative step sizes, and we are only updating step attribute when value changes (not when enabled range changes). I will fix this, but I doubt these are fully what is causing the issues with VO. Ill also run the elements through an HTML validator, that may give us more info.

**Comment by jessegreenberg** [comment]
When I run a PhET accessible slider through the validator, I see these (abbreviated)
Warning: slider role unnecessary for input of range type.
Error: Missing required attribute aria-valuenow
Error: Missing required attribute aria-valuemin
Error: Missing required attribute aria-valuemax

My guess is that the errors are because we have the role on the element, without it those aria attributes are not necessary.

I think we can remove the slider role, which was added as a workaround for a Firefox bug: https://github.com/phetsims/a11y-research/issues/51. We should see if that has been fixed.

**Comment by jessegreenberg** [comment]
When I remove the role attribute, I hear the aria-valuetext in NVDA correctly. So I removed it and all warnings and errors in the validator are gone.

@mbarlow12 can you please try master with VO again and see if this issue has been resolved?

**Comment by mbarlow12** [comment]
Just tested in RIAW and I only hear the alerts when interacting with the slider. VO properly reads out the accessible name and aria-valuetext on focus, but there's no information read out after that point.

@jessegreenberg I'm also noticing that the step is consistently being set (at least in RIAW) to exactly 0.01 below the aria-valuetext readout. I experimented with setting the 'step' attribute to the `_shiftKeyboardStep` property in `AccessibleSlider` and had appropriate output from VO, so I'm pretty sure that VO doesn't work with the variable step size. If we need that for other screen readers, then I think we may need to do some browser checking prior to handling the step size.

**Comment by jessegreenberg** [comment]
Hmm, OK thanks for checking.

Can you please clarify what you mean by this?
> I'm also noticing that the step is consistently being set (at least in RIAW) to exactly 0.01 below the aria-valuetext readout.

In RIAW I see he step change with the value to be value - enabledRange.min. Here it is with a value of 0.44.
![capture](https://user-images.githubusercontent.com/6396244/49088598-9909c200-f227-11e8-8606-dc566de39c4f.PNG)

**Comment by jessegreenberg** [comment]
Actually would it be OK to just set the step to the shift keyboard step like you tested?

**Comment by jessegreenberg** [comment]
I think this will work if we can agree it is OK for the range of the slider to always be evenly divisible by the shift step size. Ill try it out NVDA, run aqua tests, then check in with @mbarlow12 to make sure this fixes it. I think it should since this was already tested in https://github.com/phetsims/sun/issues/413#issuecomment-442075066

**Comment by jessegreenberg** [comment]
This won't work because we are counting on setting the shift keyboard step to zero in some cases.

**Comment by jessegreenberg** [comment]
@mbarlow12 and I met to work on this together. We ended up changing the solution in https://github.com/phetsims/sun/commit/6084991dd28bf12197e7b80dc251b92d62ae6d7b by setting the step size to the number of decimal places in the accessible readout. For example, if decimal places is zero, step size is 1, and if decimal places is 2, step size is 0.01. This seems to work very well on NVDA with FF, JAWS with Chrome, and VO with Safari.

This also has the benefit that step doesn't need to be adjusted every time the value changes.

Adding the "pending testing" label back, since this needs to be included in the RIAW sonification release.

**Comment by jessegreenberg** [comment]
I should mention that we considered a couple of other options too.

We considered changing the shiftKeyboardStep to not allow 0, and use shiftKeyboardStep as the value for the 'step' attribute. But this is limiting and would require changes to InteractionStateProperty, PushButtonModel, and NumberControl to support the ability to make buttons look pressed when interacting with a keyboard. This is currently supported by actually clicking the arrow buttons in NumberControl when shift key is down.

We decided against a platform specific check because the discovered solution seems to work well everywhere.

**Comment by jessegreenberg** [comment]
This fix is causing issues in https://github.com/phetsims/coulombs-law/issues/90. And on reflection, I think that accessibleDecimalPlaces should either only impact the readout of the AT (its initial intention), or become a required param. Its current usage makes the slider not work in cases like https://github.com/phetsims/coulombs-law/issues/90.

I would like to re-visit the `any` value for accessible step attribute. I think it is really what we need for PhET sliders.

**Comment by jessegreenberg** [comment]
Documenting results with step: 'any'
- NVDA + Firefox sounds great
- Firefox alone works great

**Comment by jessegreenberg** [comment]
This hasn't shown up in https://github.com/phetsims/QA/issues/247 or https://github.com/phetsims/QA/issues/235 so this can be closed now.
